File: courses/codes/cclv1-sr-class/greedy_algorithms.py
from fractions import Fraction

def egyptian_decomp():
    arr = input().split(" ")
    a, b = int(arr[0]), int(arr[1])
    denoms, q = [], 1
    
    while a > 0:
        if a == 1:
            denoms.append(b)
            break
        while 1/q > a/b:
            q += 1
        x = Fraction(a/b - 1/q).limit_denominator()
        a, b = x.numerator, x.denominator
        denoms.append(q)
    print(' '.join([str(d) for d in denoms]))



def check_conflict(s, e, A):
    for i in range(s, e+1):
        if A[i] == 1:
            return True, A
    
    for i in range(s, e+1):
        A[i] = 1

    return False, A

def scheduling():
    N = int(input())
    jobs = []
    for i in range(N):
        arr = input().split(" ")
        jobs.append((int(arr[0]), int(arr[1]), int(arr[2])))

    jobs.sort(key = lambda x: x[2], reverse=True)
    plans, accepted_jobs = [0]*11, []

    for job in jobs:
        conflict, plans = check_conflict(job[0], job[1], plans)
        if not conflict:
            accepted_jobs.append(job)

    print(sum([j[2] for j in accepted_jobs]))

# assume canonical system
def change_making():
    A = int(input())     # amount
    coins = list(map(int, input().split(" ")))
    coins.sort(reverse=True)

    count = 0
    for coin in coins:
        if A >= coin:
            qty = A // coin
            A -= qty * coin
            count += qty

    print(count)




# scheduling()
# egyptian_decomp()
# change_making()
        

File: courses/codes/cclv1-sr-class/GreedyAlgorithms.java

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

public class GreedyAlgorithms {

    public static void egyptianDecomp() {
        Scanner scanner = new Scanner(System.in);
        String[] arr = scanner.nextLine().split(" ");
        int a = Integer.parseInt(arr[0]);
        int b = Integer.parseInt(arr[1]);
        List<Integer> denoms = new ArrayList<>();
        int q = 1;

        while (a > 0) {
            if (a == 1) {
                denoms.add(b);
                break;
            }
            while (1.0 / q > (double) a / b) {
                q++;
            }
            int num = a * q - b;
            int denom = b * q;
            int gcd = gcd(num, denom);
            num /= gcd;
            denom /= gcd;
            a = num;
            b = denom;
            denoms.add(q);
        }
        for (int d : denoms) {
            System.out.print(d + " ");
        }
    }

    public static int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    public static void checkConflict(int s, int e, int[] A) {
        for (int i = s; i <= e; i++) {
            if (A[i] == 1) {
                return;
            }
        }
        for (int i = s; i <= e; i++) {
            A[i] = 1;
        }
    }

    public static void scheduling() {
        Scanner scanner = new Scanner(System.in);
        int N = Integer.parseInt(scanner.nextLine());
        int[][] jobs = new int[N][3];
        for (int i = 0; i < N; i++) {
            String[] arr = scanner.nextLine().split(" ");
            for (int j = 0; j < 3; j++) {
                jobs[i][j] = Integer.parseInt(arr[j]);
            }
        }
        Arrays.sort(jobs, (a, b) -> Integer.compare(b[2], a[2]));
        int[] plans = new int[11];
        List<int[]> acceptedJobs = new ArrayList<>();

        for (int[] job : jobs) {
            boolean conflict = false;
            for (int i = job[0]; i <= job[1]; i++) {
                if (plans[i] == 1) {
                    conflict = true;
                    break;
                }
            }
            if (!conflict) {
                acceptedJobs.add(job);
                for (int i = job[0]; i <= job[1]; i++) {
                    plans[i] = 1;
                }
            }
        }

        int total = acceptedJobs.stream().mapToInt(job -> job[2]).sum();
        System.out.println(total);
    }

    public static void changeMaking() {
        Scanner scanner = new Scanner(System.in);
        int A = Integer.parseInt(scanner.nextLine());
        String[] coinsStr = scanner.nextLine().split(" ");
        int[] coins = new int[coinsStr.length];
        for (int i = 0; i < coinsStr.length; i++) {
            coins[i] = Integer.parseInt(coinsStr[i]);
        }
        Arrays.sort(coins);
        int count = 0;
        for (int i = coins.length - 1; i >= 0; i--) {
            if (A >= coins[i]) {
                int qty = A / coins[i];
                A -= qty * coins[i];
                count += qty;
            }
        }
        System.out.println(count);
    }

    public static void main(String[] args) {
        // Uncomment the method calls for testing each function
//         scheduling();
//         egyptianDecomp();
//         changeMaking();
    }
}
